File: assign1/src/matrixproduct.java
import java.util.Arrays;
import static java.lang.Math.min;

public class matrixproduct {

    public static void OnMult(int m_ar, int m_br) {
        long time1, time2;

        double[] pha = new double[m_ar*m_ar];
        double[] phb = new double[m_ar*m_br];
        double[] phc = new double[m_ar*m_br];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i*m_ar+j] = (double)1.0;

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_br; j++)
                phb[i*m_br+j] = (double)(i + 1);

        time1 = System.currentTimeMillis();

        for (int i = 0; i < m_ar; i++) {
            for (int j = 0; j < m_br; j++) {
                double temp = 0;
                for (int k = 0; k < m_ar; k++) {
                    temp += pha[i*m_ar+k] * phb[k*m_br+j];
                }
                phc[i*m_ar+j] = temp;
            }
        }

        time2 = System.currentTimeMillis();
        double seconds = (time2 - time1) / 1000.0;

        System.out.printf("Time: %.3f seconds\n", seconds);

        System.out.println("Result matrix: ");
        for (int i = 0; i < 1; i++) {
            for (int j = 0; j < Math.min(10, m_br); j++) {
                System.out.print(phc[j] + " ");
            }
        }
        System.out.println();

    }

    public static void OnMultLine(int m_ar, int m_br) {
        long time1, time2;

        double[] pha = new double[m_ar * m_ar];
        double[] phb = new double[m_br * m_br];
        double[] phc = new double[m_ar * m_ar];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i * m_ar + j] = 1.0;

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_br; j++)
                phb[i * m_br + j] = i + 1;

        time1 = System.currentTimeMillis();

        for (int i = 0; i < m_ar; i++) {
            for (int j = 0; j < m_ar; j++) {
                for (int k = 0; k < m_ar; k++) {
                    phc[i*m_ar+k] += pha[i*m_ar+k]*phb[j*m_ar+k];
                }
            }
        }

        time2 = System.currentTimeMillis();
        double seconds = (time2 - time1) / 1000.0;

        System.out.printf("Time: %.3f seconds\n", seconds);

        System.out.println("Result matrix: ");
        for (int i = 0; i < 1; i++) {
            for (int j = 0; j < Math.min(10, m_br); j++) {
                System.out.print(phc[j] + " ");
            }
        }
        System.out.println();
    }

    public static void OnMultBlock(int m_ar, int m_br, int bkSize) {
        long time1, time2;

        double[] pha = new double[m_ar * m_ar];
        double[] phb = new double[m_br * m_br];
        double[] phc = new double[m_ar * m_ar];

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_ar; j++)
                pha[i * m_ar + j] = 1.0;

        for (int i = 0; i < m_ar; i++)
            for (int j = 0; j < m_br; j++)
                phb[i * m_br + j] = i + 1;

        time1 = System.currentTimeMillis();

        for (int ib = 0; ib<m_ar;ib+=bkSize)
			for (int jb = 0; jb<m_ar;jb+=bkSize)
				for (int kb =0;kb<m_ar;kb+=bkSize)
					for (int i =ib;i<min(ib+bkSize,m_ar);i++)
						for (int j =jb;j<min(jb+bkSize,m_ar);j++)
							for (int k=kb;k<min(kb+bkSize,m_ar);k++)
								phc[i*m_ar+k] += pha[i*m_ar+k]*phb[j*m_ar+k];

        time2 = System.currentTimeMillis();
        double seconds = (time2 - time1) / 1000.0;

        System.out.printf("Time: %.3f seconds\n", seconds);

        System.out.println("Result matrix: ");
        for (int i = 0; i < 1; i++) {
            for (int j = 0; j < Math.min(10, m_br); j++) {
                System.out.print(phc[j] + " ");
            }
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int m_ar = 3000;
        int m_br = 600;
        int bkSize = 200;
        //OnMult(m_ar, m_ar);
        OnMultLine(m_ar, m_ar);
        //OnMultBlock(m_ar, m_br, bkSize);
    }
}

File: assign1/src/matrixproduct.cpp
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <time.h>
#include <cstdlib>
#include <papi.h>
#include <omp.h>
#include <sys/time.h>

using namespace std;

#define SYSTEMTIME clock_t

 
void OnMult(int m_ar, int m_br) 
{
	
	SYSTEMTIME Time1, Time2;
	
	char st[100];
	double temp;
	int i, j, k;

	double *pha, *phb, *phc;
	

		
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

	for(i=0; i<m_ar; i++)
		for(j=0; j<m_ar; j++)
			pha[i*m_ar + j] = (double)1.0;



	for(i=0; i<m_br; i++)
		for(j=0; j<m_br; j++)
			phb[i*m_br + j] = (double)(i+1);



    Time1 = clock();

	for(i=0; i<m_ar; i++)
	{	for( j=0; j<m_br; j++)
		{	temp = 0;
			for( k=0; k<m_ar; k++)
			{	
				temp += pha[i*m_ar+k] * phb[k*m_br+j];
			}
			phc[i*m_ar+j]=temp;
		}
	}


    Time2 = clock();
	sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
	cout << st;

	// display 10 elements of the result matrix tto verify correctness
	cout << "Result matrix: " << endl;
	for(i=0; i<1; i++)
	{	for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
	}
	cout << endl;

    free(pha);
    free(phb);
    free(phc);
	
	
}

// add code here for line x line matriz multiplication
void OnMultLine(int m_ar, int m_br)
{
    SYSTEMTIME Time1, Time2;
    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    // Preencher as matrizes pha e phb com valores de exemplo
    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = double(i+1);

    Time1 = clock();

    for (i = 0; i < m_ar; i++) {
        for (j = 0; j < m_br; j++) {
            for (k = 0; k < m_ar; k++) {
                phc[i*m_ar+k] += pha[i*m_ar+k]*phb[j*m_ar+k];
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    // Exibir 10 elementos da matriz resultante para verificar a correção
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++) {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    
}
void OnMultLinePragmaOption1(int m_ar, int m_br)
{
    struct timeval start, end;
    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    // Preencher as matrizes pha e phb com valores de exemplo
    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = double(i+1);

    gettimeofday(&start,NULL);

    #pragma omp parallel for
    for (i = 0; i < m_ar; i++) {
        for (j = 0; j < m_br; j++) {
            for (k = 0; k < m_ar; k++) {
                phc[i*m_ar+k] += pha[i*m_ar+k]*phb[j*m_ar+k];
            }
        }
    }

    gettimeofday(&end,NULL);
    sprintf(st, "Time: %3.10f seconds\n", (double)((end.tv_sec - start.tv_sec)*1000000u + end.tv_usec - start.tv_usec)/1.e6);
    cout << st;

    // Exibir 10 elementos da matriz resultante para verificar a correção
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++) {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    
}
void OnMultLinePragmaOption2(int m_ar, int m_br)
{
    
    struct timeval start, end;
    char st[100];
    double temp;
    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    // Preencher as matrizes pha e phb com valores de exemplo
    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = double(i+1);

    gettimeofday(&start,NULL);

    #pragma omp parallel private(i,j)
    for (i = 0; i < m_ar; i++) {
        for (j = 0; j < m_br; j++) {
            #pragma omp for
            for (k = 0; k < m_ar; k++) {
                phc[i*m_ar+k] += pha[i*m_ar+k]*phb[j*m_ar+k];
            }
        }
    }

    gettimeofday(&end,NULL);
    sprintf(st, "Time: %3.10f seconds\n", (double)((end.tv_sec - start.tv_sec)*1000000u + end.tv_usec - start.tv_usec)/1.e6);
    cout << st;

    // Exibir 10 elementos da matriz resultante para verificar a correção
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++) {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    
}

// add code here for block x block matriz multiplication
void OnMultBlock(int m_ar, int m_br, int bkSize)
{
    SYSTEMTIME Time1, Time2;
    char st[100];
    double temp;
    int i, j, k, i_blk, j_blk, k_blk;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    // Preencher as matrizes pha e phb com valores de exemplo
    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = double(i+1);
	// ===

    Time1 = clock();

    for (i = 0; i < m_ar; i += bkSize) {
        for (j = 0; j < m_br; j += bkSize) {
            for (k = 0; k < m_ar; k += bkSize) {
                for (i_blk = i; i_blk < i + bkSize; i_blk++) {
                    for (j_blk = j; j_blk < j+ bkSize; j_blk++) {
                        for (k_blk = k; k_blk < k+bkSize; k_blk++) {
                            phc[i_blk*m_ar+k_blk] += pha[i_blk * m_ar + j_blk] * phb[j_blk * m_br + k_blk];
                        }
                    }
                }
            }
        }
    }

    Time2 = clock();
    sprintf(st, "Time: %3.3f seconds\n", (double)(Time2 - Time1) / CLOCKS_PER_SEC);
    cout << st;

    // Exibir 10 elementos da matriz resultante para verificar a correção
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++) {
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    
}



void handle_error (int retval)
{
  printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
  exit(1);
}

void init_papi() {
  int retval = PAPI_library_init(PAPI_VER_CURRENT);
  if (retval != PAPI_VER_CURRENT && retval < 0) {
    printf("PAPI library version mismatch!\n");
    exit(1);
  }
  if (retval < 0) handle_error(retval);

  std::cout << "PAPI Version Number: MAJOR: " << PAPI_VERSION_MAJOR(retval)
            << " MINOR: " << PAPI_VERSION_MINOR(retval)
            << " REVISION: " << PAPI_VERSION_REVISION(retval) << "\n";
}


int main (int argc, char *argv[])
{

	char c;
	int lin, col, blockSize;
	int op;
	
	int EventSet = PAPI_NULL;
  	long long values[3];
  	int ret;
	

	ret = PAPI_library_init( PAPI_VER_CURRENT );
	if ( ret != PAPI_VER_CURRENT )
		std::cout << "FAIL" << endl;


	ret = PAPI_create_eventset(&EventSet);
		if (ret != PAPI_OK) cout << "ERROR: create eventset" << endl;


	ret = PAPI_add_event(EventSet,PAPI_L1_DCM );
	if (ret != PAPI_OK) cout << "ERROR: PAPI_L1_DCM" << endl;


	ret = PAPI_add_event(EventSet,PAPI_L2_DCM);
	if (ret != PAPI_OK) cout << "ERROR: PAPI_L2_DCM" << endl;

    ret = PAPI_add_event(EventSet,PAPI_DP_OPS);
	if (ret != PAPI_OK) cout << "ERROR: PAPI_DP_OPS" << endl;


	op=1;
	do {
		cout << endl << "1. Multiplication" << endl;
		cout << "2. Line Multiplication" << endl;
		cout << "3. Block Multiplication" << endl;
        cout << "4. Line Multiplication - Pragma Option 1" << endl;
        cout << "5. Line Multiplication - Pragma Option 2" << endl;
		cout << "Selection?: ";
		cin >>op;
		if (op == 0)
			break;
		printf("Dimensions: lins=cols ? ");
   		cin >> lin;
   		col = lin;


		// Start counting
		ret = PAPI_start(EventSet);
		if (ret != PAPI_OK) cout << "ERROR: Start PAPI" << endl;

		switch (op){
			case 1: 
				OnMult(lin, col);
				break;
			case 2:
				OnMultLine(lin, col);  
				break;
			case 3:
				cout << "Block Size? ";
				cin >> blockSize;
				OnMultBlock(lin, col, blockSize);  
				break;
            case 4:
				OnMultLinePragmaOption1(lin, col);  
				break;
            case 5:
				OnMultLinePragmaOption2(lin, col);  
				break;
		}

  		ret = PAPI_stop(EventSet, values);
  		if (ret != PAPI_OK) cout << "ERROR: Stop PAPI" << endl;
  		printf("L1 DCM: %lld \n",values[0]);
  		printf("L2 DCM: %lld \n",values[1]);
        printf("FLOPS: %lld \n",values[2]);

		ret = PAPI_reset( EventSet );
		if ( ret != PAPI_OK )
			std::cout << "FAIL reset" << endl; 



	}while (op != 0);

	ret = PAPI_remove_event( EventSet, PAPI_L1_DCM );
	if ( ret != PAPI_OK )
		std::cout << "FAIL remove event" << endl; 

	ret = PAPI_remove_event( EventSet, PAPI_L2_DCM );
	if ( ret != PAPI_OK )
		std::cout << "FAIL remove event" << endl; 

    ret = PAPI_remove_event( EventSet, PAPI_DP_OPS );
	if ( ret != PAPI_OK )
		std::cout << "FAIL remove event" << endl;

	ret = PAPI_destroy_eventset( &EventSet );
	if ( ret != PAPI_OK )
		std::cout << "FAIL destroy" << endl;

}